x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -383,7 +385,8 @@ protected LogicalPlan rule(Limit limit) {
                 // check if there's a 'visible' descendant limit lower than the current one
                 // and if so, align the current limit since it adds no value
                 // this applies for cases such as | limit 1 | sort field | limit 10
-                else {
+                // but NOT for mv_expand (ie | limit 1 | mv_expand x | limit 20) where we want that last "limit" to apply on expand results
+                else if (unary instanceof MvExpand == false) {
```

Comment:
Don't combine the limits for mv_expand.

Reply:
this works only if the limit is right above `mv_expand` - if it's hidden by another node (keep, where, etc..) the rule won't see it.
Move the check into descendantLimit so that if mv_expand is encountered, just like with an agg, no limit is returned.

Reply:
Good point. Haven't thought of this scenario.